perf: constant time checkpointing

agostbiro opened the pull request:
In https://github.com/NomicFoundation/edr/issues/212 we established that we spend most of the time checkpointing state while executing Hardhat test suites. This PR changes the previous O(n) checkpointing behavior to O(1) by using persistent data structures for the state.

Note that contrary to previous discussions I didn't change the LRU cache in `ProviderData` to a vector as the LRU cache overhead didn't show up while profiling and it's a good option to have a limit on the size of the state in memory even if it's much smaller now. I did change however the block number to state id map to a persistent one as cloning this for snapshots showed up in performance profiles.

agostbiro commented on crates/tools/src/main.rs:
```diff
@@ -11,6 +11,10 @@ mod update;
 
+// Matches `edr_napi`. Important for scenarios.
+#[global_allocator]
+static ALLOC: mimalloc::MiMalloc = mimalloc::MiMalloc;
```
Added this to match allocator used by `edr_napi` when executing scenarios from Rust

agostbiro commented:
CI is fixed now, please see [Reduce max cached states](https://github.com/NomicFoundation/hardhat/pull/5136/commits/688c1692b24a1cc584dcee4c9635ed1f5f183ef0) for explanation

Wodann commented on crates/edr_evm/src/state/trie/account.rs:
```diff
@@ -1,317 +1,93 @@
 }
 
-impl AccountTrie {
+impl<'a> AccountTrie {
```
Request for change:

The lifetime doesn't have to be on the `impl` block, as it's not used on the type. You can have `<'a>` on the individual functions that need them

agostbiro replied:
Thanks, it was a left-over from a previous iteration. Fixed now

Wodann commented on crates/edr_evm/src/state/trie/account.rs:
```diff
@@ -399,153 +150,167 @@ impl AccountTrie {
-                storage_root.as_slice(),
-            )
-            .expect("Invalid storage root");
+    fn mutate(&'a mut self) -> AccountTrieMutation<'a> {
```
It might even suffice here to do:
```suggestion
    fn mutate(&mut self) -> AccountTrieMutation<'_> {
```

agostbiro replied:
Fixed

Wodann commented on crates/edr_evm/src/state/trie/account.rs:
```diff
@@ -399,153 +150,167 @@ impl AccountTrie {
+        );
+    }
+
+    /// Create or update teh account info. Ensures that a storage trie exists.
```
```suggestion
    /// Create or update the account info. Ensures that a storage trie exists.
```

Wodann commented on crates/edr_evm/src/state/trie/state_trie.rs:
```diff
@@ -0,0 +1,98 @@
+        let mut trie = TrieQuery::empty(Arc::clone(&db));
+        let root = trie.root();
+
+        // TODO root should be cacheable
```
Should this be handled in this PR? If not, could you please create a GitHub issue for it and link it?

Wodann reviewed (changes requested):
Nicely done. I like the structure of the helper structs for mutations.

I left comments where necessary, but in general I'd like to see some tests to validate the persistent storage. Currently, there are no tests validating that behaviour

agostbiro commented on .changeset/tough-students-compare.md:
```diff
@@ -0,0 +1,5 @@
+---
+@nomicfoundation/edr: minor
```
Thanks, the missing quotes issue was fixed. 

Re minor change: I prefixed the commit with "feat", because it felt that it better communicated the change than "fix" as there was nothing broken, we just changed to a better approach. And "feat" implies "minor" bump in the [semver](https://www.conventionalcommits.org/en/v1.0.0/) spec.

In any case, I'm happy to change it to "fix" and "patch".

Wdyt @fvictorio ?

agostbiro replied:
Ok thanks, changed to patch. Ok to keep "feat" in the prefix?

fvictorio replied:
Yes, I'm ok with `feat`. Alternatively we can use a `perf` prefix. I don't have a preference.

Wodann replied:
If we're using the conventional commit standard for prefixes, this would be a `perf:`

agostbiro commented on crates/edr_evm/src/collections/shared_map.rs:
```diff
@@ -1,14 +1,14 @@
 }
 
-impl<T> SharedMapEntry<T> {
+impl<T: Clone> SharedMapEntry<T> {
```
It doesn't actually, thanks fixed it.

agostbiro commented on crates/edr_evm/src/state/trie/storage_trie.rs:
```diff
@@ -0,0 +1,118 @@
+    fn default() -> Self {
+        let db = Arc::new(PersistentMemoryDB::default());
+        let mut trie = TrieQuery::empty(Arc::clone(&db));
+        // TODO root should be cacheable
```
I thought this was a nice to have, but ran out of time. Imo we should only do this in the future if it shows up on a performance profile, so I just removed the comment and not adding an issue for it.

agostbiro commented on crates/edr_evm/src/state/trie/storage_trie.rs:
```diff
@@ -0,0 +1,118 @@
+        TrieQuery::new(Arc::clone(&self.db), &self.root)
+    }
+
+    fn decode_u256(encoded_value: Vec<u8>) -> U256 {
+        U256::decode(&mut encoded_value.as_slice()).expect("Valid RLP")
+    }
```
Happy to change this, just fyi: I did it this way to follow the previous pattern in this module where associated functions not using self were added as static methods.

Wodann replied:
I don't know that I wrote any functions as associated methods if they didn't use `self` or `Self`; hence why I left the comment.

agostbiro commented on crates/edr_evm/src/state/trie/storage_trie.rs:
```diff
@@ -0,0 +1,118 @@
+};
+
+#[derive(Debug)]
+pub(super) struct StorageTrie {
```
Respectfully disagree: This struct is only exposed to the parent module which is why the lint checking for missing docs doesn't flag this. If there are new requirements on docs coverage, that should be a separate task imo.

fvictorio commented:
This one goes over my head so Remco's approval should be enough.

agostbiro commented on crates/edr_evm/src/state/trie/storage_trie.rs:
```diff
@@ -0,0 +1,118 @@
+}
+
+impl<'a> StorageTrie {
+    pub fn mutate(&'a mut self) -> StorageTrieMutation<'a> {
```
I think this is slightly different. `get_mut` is for things that wrap a type to get a mutable reference for the underlying type and `as_mut` is typically used to support cases where the underlying data can be mutated through different interfaces (e.g. as mut slice or as mut vec).

The mutation structs in this change serve a different purpose: they ensure that the state root is updated after a series of mutations and that stale state roots cannot be read while mutations are being applied.

agostbiro commented on crates/edr_evm/src/state/trie/trie_query.rs:
```diff
@@ -0,0 +1,79 @@
+            .expect(DB_IS_INFALLIBLE);
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (Vec<u8>, Vec<u8>)> + '_ {
```
Added

agostbiro commented on crates/edr_evm/src/state/trie/account.rs:
```diff
@@ -1,317 +1,93 @@
+                    // TODO question to reviewers: does it come from the Ethereum protocol or is it
+                    // a quirk of our implementation that an account can already have a storage trie
+                    // when it's status is created?
                     if account.is_created() {
```
Thanks! Added a link to the source

agostbiro commented:
Thanks for the review! I addressed comments in [Incorporate review feedback](https://github.com/NomicFoundation/hardhat/pull/5136/commits/009ed7c91aeba664447509a65202834046223ece) and [added clone test for TrieState](https://github.com/NomicFoundation/hardhat/pull/5136/commits/d7815bfc529a9d99fb3e4e895ee4fc62649cc468) as requested.
